client/fingerprint_manager.go

```diff
@@ -73,43 +83,149 @@ func (fm *FingerprintManager) setupDrivers(drivers []string) error {
 			return err
 		}
 
-		detected, err := fm.fingerprint(name, d)
+		detected, err := fm.fingerprintDriver(name, d)
 		if err != nil {
 			fm.logger.Printf("[DEBUG] client.fingerprint_manager: fingerprinting for %v failed: %+v", name, err)
 			return err
 		}
 
+		if hc, ok := d.(fingerprint.HealthCheck); ok {
+			fm.runHealthCheck(name, hc)
+		} else {
+			// for drivers which are not of type health check, update them to have their
+			// health status match the status of whether they are detected or not.
+			healthInfo := &structs.DriverInfo{
+				Healthy:    detected,
+				UpdateTime: time.Now(),
+			}
+			if node := fm.updateNodeFromDriver(name, nil, healthInfo); node != nil {
+				fm.nodeLock.Lock()
+				fm.node = node
+				fm.nodeLock.Unlock()
+			}
+		}
+
+		go fm.watchDriver(d, name)
+
 		// log the fingerprinters which have been applied
 		if detected {
 			availDrivers = append(availDrivers, name)
 		}
-
-		p, period := d.Periodic()
-		if p {
-			go fm.run(d, period, name)
-		}
 	}
 
 	fm.logger.Printf("[DEBUG] client.fingerprint_manager: detected drivers %v", availDrivers)
 	return nil
 }
 
+// watchDrivers facilitates the different periods between fingerprint and
+// health checking a driver
+func (fm *FingerprintManager) watchDriver(d driver.Driver, name string) {
+	_, fingerprintPeriod := d.Periodic()
+	fm.logger.Printf("[DEBUG] client.fingerprint_manager: fingerprinting driver %s every %v", name, fingerprintPeriod)
+
+	var healthCheckPeriod time.Duration
+	hc, isHealthCheck := d.(fingerprint.HealthCheck)
+	if isHealthCheck {
+		req := &cstructs.HealthCheckIntervalRequest{}
+		var resp cstructs.HealthCheckIntervalResponse
+		hc.GetHealthCheckInterval(req, &resp)
+		if resp.Eligible {
+			fm.logger.Printf("[DEBUG] client.fingerprint_manager: health checking driver %s every %v", name, healthCheckPeriod)
+			healthCheckPeriod = resp.Period
+		}
+	}
+
+	t1 := time.NewTimer(fingerprintPeriod)
```

Comment:
In this case you may want to use a ticker so you don't have to reset: https://golang.org/pkg/time/#NewTicker

Reply:
Can you put better names for the timers